src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SDSEarlyReleaseDefaultingRulesService.kt

```diff
@@ -29,144 +29,198 @@ class SDSEarlyReleaseDefaultingRulesService(
+    adjustDates(
+      earlyReleaseResult,
       standardReleaseResult,
+      trancheCommencementDate,
+      allocatedTranche,
+      originalBooking,
+      dates,
       breakdownByReleaseDateType,
     )
 
-    handleCRDorARDAndPRRD(dates, earlyReleaseResult.otherDates.toMutableMap(), breakdownByReleaseDateType, standardReleaseResult)
+    return createFinalCalculationResult(
+      dates,
+      breakdownByReleaseDateType,
+      earlyReleaseResult,
+      standardReleaseResult,
+      allocatedTranche,
+    )
+  }
+
+  private fun adjustDates(
+    earlyReleaseResult: CalculationResult,
+    standardReleaseResult: CalculationResult,
+    trancheCommencementDate: LocalDate?,
+    allocatedTranche: SDSEarlyReleaseTranche,
+    originalBooking: Booking,
+    dates: MutableMap<ReleaseDateType, LocalDate>,
+    breakdownByReleaseDateType: MutableMap<ReleaseDateType, ReleaseDateCalculationBreakdown>,
+  ) {
+    adjustDatesForEarlyOrStandardRelease(
+      earlyReleaseResult,
+      standardReleaseResult,
+      trancheCommencementDate,
+      allocatedTranche,
+      dates,
+      breakdownByReleaseDateType,
+    )
 
-    handleCRDEqualsEligibilityDateAndTrancheDate(dates)
+    adjustTusedForPreTrancheOneRecalls(
+      dates,
+      originalBooking,
+      standardReleaseResult,
+      breakdownByReleaseDateType,
+    )
 
-    return CalculationResult(
+    adjustHdcedAndPrrdBasedOnCrdOrArd(
       dates,
+      earlyReleaseResult.otherDates.toMutableMap(),
       breakdownByReleaseDateType,
-      earlyReleaseResult.otherDates,
-      earlyReleaseResult.effectiveSentenceLength,
-      sdsEarlyReleaseAllocatedTranche = allocatedTranche,
-      sdsEarlyReleaseTranche = allocatedTranche,
-      affectedBySds40 = (dates != standardReleaseResult.dates),
+      standardReleaseResult,
     )
+
+    removeHdcedAndErsedIfMatchingRelease(dates)
   }
 
-  private fun getControllingDate(
+  private fun createFinalCalculationResult(
+    dates: Map<ReleaseDateType, LocalDate>,
+    breakdownByReleaseDateType: Map<ReleaseDateType, ReleaseDateCalculationBreakdown>,
+    earlyReleaseResult: CalculationResult,
+    standardReleaseResult: CalculationResult,
+    allocatedTranche: SDSEarlyReleaseTranche,
+  ): CalculationResult = CalculationResult(
+    dates = dates,
+    breakdownByReleaseDateType = breakdownByReleaseDateType,
+    otherDates = earlyReleaseResult.otherDates,
+    effectiveSentenceLength = earlyReleaseResult.effectiveSentenceLength,
+    sdsEarlyReleaseAllocatedTranche = allocatedTranche,
+    sdsEarlyReleaseTranche = allocatedTranche,
+    affectedBySds40 = (dates != standardReleaseResult.dates),
+  )
+
+  private fun getArdOrCrd(
     dates: MutableMap<ReleaseDateType, LocalDate>,
-  ): Pair<ReleaseDateType, LocalDate?> {
-    return if (dates[ReleaseDateType.ARD] != null) ReleaseDateType.ARD to dates[ReleaseDateType.ARD] else ReleaseDateType.CRD to dates[ReleaseDateType.CRD]
-  }
+  ): Pair<ReleaseDateType, LocalDate?> =
+    when {
```

Comment:
This is great. You have to catch all possible outcomes with this approach and its easier to read. Especially with Sealed Classes.